resolwe/flow/tests/test_api.py

```diff
@@ -367,6 +370,89 @@ def test_change_collection(self):
         data.refresh_from_db()
         self.assertEqual(data.collection, self.collection)
 
+    def test_move_to_collection(self):
+        data_in_entity = Data.objects.create(
+            contributor=self.contributor, process=self.proc
+        )
+        data_orphan = Data.objects.create(
+            contributor=self.contributor, process=self.proc
+        )
+        data_orphan.entity.data.remove(data_orphan)
+
+        collection_1 = Collection.objects.create(contributor=self.contributor)
+        collection_1.data.add(data_orphan)
+        assign_perm("view_collection", self.contributor, collection_1)
+        assign_perm("edit_collection", self.contributor, collection_1)
+
+        collection_2 = Collection.objects.create(contributor=self.contributor)
+        assign_perm("view_collection", self.contributor, collection_2)
+        assign_perm("edit_collection", self.contributor, collection_2)
+
+        # Assert preventing moving data in entity.
+        request = factory.post(
+            reverse("resolwe-api:data-move-to-collection"),
+            {"ids": [data_in_entity.id], "destination_collection": collection_2.id,},
+            format="json",
+        )
+        force_authenticate(request, self.contributor)
+        response = self.move_to_collection_viewset(request)
+
+        self.assertEqual(
+            response.data["error"],
+            "If Data is in entity, you can only move it to another collection by moving entire entity.",
+        )
+
+        # Assert moving data not in entity.
+        request = factory.post(
+            reverse("resolwe-api:data-move-to-collection"),
+            {"ids": [data_orphan.id], "destination_collection": collection_2.id,},
+            format="json",
+        )
+
+        self.assertEqual(collection_1.data.count(), 1)
+        self.assertEqual(collection_2.data.count(), 0)
+
+        force_authenticate(request, self.contributor)
+        response = self.move_to_collection_viewset(request)
+
+        self.assertEqual(collection_1.data.count(), 0)
+        self.assertEqual(collection_2.data.count(), 1)
+
+        # Assert preventing moving data if destination collection
+        # lacks permissions.
+        remove_perm("edit_collection", self.contributor, collection_1)
+        request = factory.post(
+            reverse("resolwe-api:data-move-to-collection"),
+            {"ids": [data_orphan.id], "destination_collection": collection_1.id,},
+            format="json",
+        )
+        force_authenticate(request, self.contributor)
+        response = self.move_to_collection_viewset(request)
+
+        self.assertEqual(
+            response.data["detail"],
+            "You do not have permission to perform this action.",
+        )
+
+        # It's possible to move data if source collection does
+        # not have add and edit permissions.
```

Comment:
Remove "add". It should not be possible to move the data if you don't have edit permission on both collections.